### PR TITLE
Add new rpmExpandThisMacro() public method

### PR DESCRIPTION
**Pull request by mlschroe**

### CODE REVIEW

**Comment by mlschroe** [comment]
This needs to be connected to python/lua as well. How should the interface look like?

**Comment by pmatilai** [comment]
I can pick up the Lua side once this is merged, no need to redo all that. As for Python, the bindings wrt macros are really primitive as it is and nobody is complaining so I dunno if anybody cares. At any rate, I don't consider that a requirement for merging.

**Comment by pmatilai** [comment]
...except that's exactly where it starts getting trickier: from Lua the argv is differently allocated and shouldn't be freed in freeArgs() and at least in my version there was "fun" with me->opts processing retriggering the option processing when it shouldn't have, etc. But many details are different here, I'll rebase the Lua work on top of this and see how it looks like now.

**Comment by pmatilai** [comment]
Ah, making a copy of the argv so the allocation business is different from what I had. https://github.com/pmatilai/rpm/tree/luamt-2 is a quick "port" on top of this and the basic testcases pass as-is, but only very briefly tested.

**Comment by pmatilai** [comment]
Oh and BTW, there's actually an existing use-case for this in the code-base: this is exactly what runCall() in build/rpmfc.c wants.

**Comment by Conan-Kudo** [comment]
@pmatilai `spectool` was rewritten to use RPM Python bindings, so I imagine at least it would matter for that tool.

**Comment by pmatilai** [comment]
> spectool was rewritten to use RPM Python bindings, so I imagine at least it would matter for that tool.

Hardly. There's precisely *one* rpm.expandMacro() in the entire new spectool code, and that doesn't involve parametric macros or anything else advanced, it's merely to expand %_sourcedir.

**Comment by mlschroe** [comment]
I wonder if we should make the interface more sane. I currently made it similar to rpmExpandMacro(), but that was probably a bad idea. The interface is currently return 1 on success and -1 on failure, put string in the obuf pointer.
But the straightforward interface would be return string on success, NULL on failure.

**Comment by pmatilai** [comment]
Heh, I almost commented on the return code thing, before noticing that it was actually to be same as rpmExpandMacro(). There was a kind of a reason for the return code to be like that, basically to leave room for returning more information in the future (but I don't recall if there was an actual plan as that what exactly).

Just had a closer look at the actual patches, my only real complaint is the amount duplication of code between expandMacro() and expandThisMacro(). That's the main difference between your version and mine, and avoids the recursion issues I ran into (so difference between working and non-working :grin: ), but... maybe some of the entry/exit mb checks could be further refactored to use common code?

**Comment by pmatilai** [comment]
Okay, just for the reference, I suggested considering using an error code instead of NULL: https://github.com/rpm-software-management/rpm/pull/34 but as to how that turned into -1/1 I plead innocent :D

The documented "return negative on error" would allow changing it to return 0 on success like the inner calls do (so you could just pass the inner rc on), I certainly would not mind...

File: rpmio/macro.c [hunk 1]
```diff
@@ -1545,33 +1581,14 @@ expandMacro(MacroBuf mb, const char *src, size_t slen)
...
-		expandMacro(mb, me->body, 0);
-	}
-
-	/* Free args for "%name " macros with opts */
-	if (args != NULL)
-	    freeArgs(mb);
-	mb->args = prevarg;
-	mb->me = prevme;
-
+	doExpandThisMacro(mb, me, args);
```
Comment (pmatilai) [inline rpmio/macro.c:1591]: Oh, the other thing is that this looks a lot like a memory leak, as the argv free is buried far away from here. I'd prefer symmetry here - args was initially allocated here so it should also be freed here, ditto with optargs in expandThisMacro().

**Comment by pmatilai** [comment]
Yet another BTW: I don't think it would be unreasonable to expect rpmExpandThisMacro() callers to supply the macro name as the first argument of the argv (similar to the exec() family of functions) so we wouldn't need to copy that internally just because - the caller knows the name and has to build the argv anyway...
With that you could technically even drop the "n" argument, but whether that'd make the interface too weird I dunno.

**Comment by mlschroe** [comment]
So you're suggesting that we remove the argvFree() call in the freeArgs() function?

**Comment by mlschroe** [comment]
I'll update the patch set after lunch.

**Comment by Conan-Kudo** [comment]
> > spectool was rewritten to use RPM Python bindings, so I imagine at least it would matter for that tool.
> 
> Hardly. There's precisely _one_ rpm.expandMacro() in the entire new spectool code, and that doesn't involve parametric macros or anything else advanced, it's merely to expand %_sourcedir.

That's true, however, it is reasonable to assume that other programs written in Python would need to do this, especially as parametric macros become more common. RPMLint does this in some places too. _I've_ written code to use `rpm.expandMacro()` for personal tools too...

**Comment by pmatilai** [comment]
@Conan-Kudo , the point is that Python and macros reside on different levels of existence entirely. In the embedded Lua realm, you *are* a macro most of the time, whereas Python only sometimes casually observes the value from the outside. Nobody is forbidding folks from adding more sophisticated macro bindings for Python but in reality the use-cases don't seem to be there.

**Comment by mlschroe** [comment]
Ok, I factored out the common code. I'm not super happy about the naming, though. Comment welcome ;)

**Comment by pmatilai** [comment]
Names are annoying... for a theme that's used elsewhere in rpm too, mbInit() and mbFini() maybe?
Also seems to me in that the mbAllocBuf() calls could be moved to the start/init-part too. But yeah much nicer this way, regardless of names.

**Comment by pmatilai** [comment]
Oh, sorry I missed that last push from yesterday. Looks fine now, except that the last commit message still talks about mbStartExpansion() and mbFinishExpansion().

**Review by pmatilai — APPROVED** [review]

**Comment by pmatilai** [comment]
Thanks!

**Pull request merged**